### PR TITLE
redo regressor trees

### PR DESCRIPTION
**Pull request by SUKI-O**
Inplement #28 
Closes: #65

Note: This is a redo of #65 which I had to scratch.  ( Will keep #65 open while being reviewed)

Changes proposed in this pull request:

- Add ObliqueDecisionTreeRegressor
- Add PatchObliqueDecisionTreeRegressor
- Add ObliqueRadomForestRegressor
- Add PatchObliqueRadomForestRegressor

- Add tests for ObliqueDecisionTreeRegressor
- Add tests for PatchObliqueDecisionTreeRegressor
- Add tests for ObliqueRadomForestRegressor
- Add test for PatchObliqueRadomForestRegressor

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [x] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by SUKI-O** [comment]
Not sure why some tests are failing. I hope I didn't mess up this time again...

File: sktree/ensemble/_supervised_forest.py [hunk 1]
```diff
@@ -335,6 +340,264 @@ def __init__(
         self.min_impurity_decrease = min_impurity_decrease
 
 
+class ObliqueRandomForestRegressor(ForestRegressor):
+    """
+    An oblique random forest regressor.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py]:
```suggestion
    An oblique random forest regressor.
    
```

File: sktree/ensemble/_supervised_forest.py [hunk 2]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    A oblique random forest is a meta estimator similar to a random
+    forest that fits a number of oblique decision tree regressor
+    on various sub-samples of the dataset and uses averaging to
+    improve the predictive accuracy and control over-fitting.
+    The sub-sample size is controlled with the `max_samples` parameter if
+    `bootstrap=True` (default), otherwise the whole dataset is used to build
+    each tree.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:355]:
```suggestion
    each tree.
    
```

File: sktree/ensemble/_supervised_forest.py [hunk 3]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    Read more in the :ref:`User Guide <sklearn:forest>`.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:357]:
```suggestion
    Read more in the :ref:`User Guide <sklearn:forest>`.

```

File: sktree/ensemble/_supervised_forest.py [hunk 4]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+          `max_samples` should be in the interval `(0.0, 1.0]`.
+    feature_combinations : float, default=None
+        The number of features to combine on average at each split
+        of the decision trees. If ``None``, then will default to the minimum of
+        ``(1.5, n_features)``. This controls the number of non-zeros is the
+        projection matrix. Setting the value to 1.0 is equivalent to a
+        traditional decision-tree. ``feature_combinations * max_features``
+        gives the number of expected non-zeros in the projection matrix of shape
+        ``(max_features, n_features)``. Thus this value must always be less than
+        ``n_features`` in order to be valid.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:492]:
```suggestion
        ``n_features`` in order to be valid.

```

File: sktree/ensemble/_supervised_forest.py [hunk 5]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    oob_score_ : float
+        Score of the training dataset obtained using an out-of-bag estimate.
+        This attribute exists only when ``oob_score`` is True.
+    oob_decision_function_ : ndarray of shape (n_samples, n_classes) or \
+            (n_samples, n_classes, n_outputs)
+        Decision function computed with out-of-bag estimate on the training
+        set. If n_estimators is small it might be possible that a data point
+        was never left out during the bootstrap. In this case,
+        `oob_decision_function_` might contain NaN. This attribute exists
+        only when ``oob_score`` is True.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:537]:
```suggestion
        only when ``oob_score`` is True.

```

File: sktree/ensemble/_supervised_forest.py [hunk 6]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    See Also
+    --------
+    sktree.tree.ObliqueDecisionTreeRegressor : An oblique decision
+        tree regressor.
+    sklearn.ensemble.RandomForestRegressor : An axis-aligned decision
+        forest regressor.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:544]:
```suggestion
        forest regressor.

```

File: sktree/ensemble/_supervised_forest.py [hunk 7]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    (e.g. ``max_depth``, ``min_samples_leaf``, etc.) lead to fully grown and
+    unpruned trees which can potentially be very large on some data sets. To
+    reduce memory consumption, the complexity and size of the trees should be
+    controlled by setting those parameter values.
+    The features are always randomly permuted at each split. Therefore,
+    the best found split may vary, even with the same training data,
+    ``max_features=n_features`` and ``bootstrap=False``, if the improvement
+    of the criterion is identical for several splits enumerated during the
+    search of the best split. To obtain a deterministic behaviour during
+    fitting, ``random_state`` has to be fixed.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:559]:
```suggestion
    fitting, ``random_state`` has to be fixed.

```

File: sktree/ensemble/_supervised_forest.py [hunk 8]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    References
+    ----------
+    .. [1] L. Breiman, "Random Forests", Machine Learning, 45(1), 5-32, 2001.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:563]:
```suggestion
    .. [1] L. Breiman, "Random Forests", Machine Learning, 45(1), 5-32, 2001.

```

Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py:563]: I would add the reference for the Sparse Projection Oblique Random Forests (tomita et al.) paper here

File: sktree/ensemble/_supervised_forest.py [hunk 9]
```diff
@@ -335,6 +340,264 @@ def __init__(
...
+    >>> from sktree.ensemble import ObliqueRandomForestRegressor
+    >>> from sklearn.datasets import make_regression
+    >>> X, y = make_regression(n_features=4, n_informative=2,
+    ...                        random_state=0, shuffle=False)
+    >>> regr = ObliqueRandomForestRegressor(max_depth=2, random_state=0)
+    >>> regr.fit(X, y)
+    ObliqueRandomForestRegressor(...)
+    >>> print(regr.predict([[0, 0, 0, 0]]))
+    # TODO: Add example output
+    # [1]
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py]: Add output?

File: sktree/ensemble/_supervised_forest.py [hunk 10]
```diff
@@ -669,3 +932,263 @@ def __init__(
         self.min_weight_fraction_leaf = min_weight_fraction_leaf
         self.max_leaf_nodes = max_leaf_nodes
         self.min_impurity_decrease = min_impurity_decrease
+
+
+class PatchObliqueRandomForestRegressor(ForestRegressor):
+    """A patch-oblique random forest regressor.
```
Comment (adam2392) [inline sktree/ensemble/_supervised_forest.py]: Can you fix the spacing in the docstring as I have done above?

File: sktree/tree/_classes.py [hunk 11]
```diff
@@ -925,59 +930,52 @@ def _build_tree(
...
-    A patch oblique decision tree is also known as a manifold oblique decision tree
-    (called MORF in :footcite:`Li2019manifold`), where the splitter is aware of
-    the structure in the data. For example, in an image, a patch would be contiguous
-    along the rows and columns of the image. In a multivariate time-series, a patch
-    would be contiguous over time, but possibly discontiguous over the sensors.
-
+class ObliqueDecisionTreeRegressor(DecisionTreeRegressor):
+    """A decision tree Regressor.
+    Read more in the :ref:`User Guide <sklearn:tree>`. The implementation follows
+    that of :footcite:`breiman2001random` and :footcite:`TomitaSPORF2020`.
```
Comment (adam2392) [inline sktree/tree/_classes.py]: Can you fix the spacing in the docstrings as I have suggested in the `_supervised_forest.py` file?

File: sktree/tree/_classes.py [hunk 12]
```diff
@@ -1005,152 +1000,117 @@ class PatchObliqueDecisionTreeClassifier(DecisionTreeClassifier):
...
-        If provided, then the feature weights are used to weight the
-        patches that are generated. The feature weights are used
-        as follows: for every patch that is sampled, the feature weights over
-        the entire patch is summed and normalizes the patch.
-
+    ccp_alpha : non-negative float, default=0.0
+        Complexity parameter used for Minimal Cost-Complexity Pruning. The
+        subtree with the largest cost complexity that is smaller than
+        ``ccp_alpha`` will be chosen. By default, no pruning is performed. See
+        :ref:`minimal_cost_complexity_pruning` for details.
```
Comment (adam2392) [inline sktree/tree/_classes.py]:
```suggestion
    ccp_alpha : non-negative float, default=0.0
        Complexity parameter used for Minimal Cost-Complexity Pruning. Oblique trees
        do not support cost complexity pruning yet. See
        :ref:`minimal_cost_complexity_pruning` for details.
```

File: sktree/tree/_classes.py [hunk 13]
```diff
@@ -1206,43 +1166,863 @@ def fit(self, X, y, sample_weight=None, check_input=True):
...
-            if issparse(X):
-                X.sort_indices()
+        n_samples, n_features = X.shape
 
-                if X.indices.dtype != np.intc or X.indptr.dtype != np.intc:
-                    raise ValueError("No support for np.int64 index based sparse matrices")
+        if self.feature_combinations is None:
+            self.feature_combinations_ = min(n_features, 1.5)
+        elif self.feature_combinations > n_features:
+            raise RuntimeError(
```
Comment (adam2392) [inline sktree/tree/_classes.py:1225]: Can you add a unit test for this line (codecov indicates it is not covered)

File: sktree/tree/_classes.py [hunk 14]
```diff
@@ -1206,43 +1166,863 @@ def fit(self, X, y, sample_weight=None, check_input=True):
...
+                self.max_features_,
+                min_samples_leaf,
+                min_weight_leaf,
+                random_state,
+                self.feature_combinations_,
+            )
+
+        if is_classifier(self):
+            raise ValueError("Please use the ObliqueDecisionTreeClassifier for classification")
+        elif is_regressor(self):
```
Comment (adam2392) [inline sktree/tree/_classes.py]: You don't need these lines actually

**Review by adam2392 — COMMENTED** [review]
Kay looks pretty good. There are a bunch of formatting issues and also a bunch of LOC that appears to be flagged by codecov as not tested.

Can you:

1. fix the formatting
2. add some unit-tests to fix the coverage
3. Add relevant entries to the documentation such as adding to the `api.rst` file (see https://output.circle-artifacts.com/output/job/8904d0d2-e37c-4030-9133-9d751b8a72a3/artifacts/0/dev/api.html where we would expect to see renderings of the docstring available)
4. Address any of my remaining comments

I am not entirely sure why the CI is failing, but it doesn't seem related to your PR. It might be something going on w/ scikit-learn upstream. We can probably move forward w/ this PR and downstream PRs while I investigate that. 

Feel free to ping me w/ questions / once you address the above. Feel free to also resolve the GH conversations if you have addressed them

**Comment by SUKI-O** [comment]
Running this` mypy -p sktree --config-file pyproject.toml` results in
`sktree/tree/tests/test_tree.py:44: error: Dict entry 0 has incompatible type "str": "Type[ObliqueDecisionTreeClassifier]"; expected "str": "Type[object]"` 
Not sure what that's about. Any thoughts?

**Comment by adam2392** [comment]
> Running this` mypy -p sktree --config-file pyproject.toml` results in
> `sktree/tree/tests/test_tree.py:44: error: Dict entry 0 has incompatible type "str": "Type[ObliqueDecisionTreeClassifier]"; expected "str": "Type[object]"`
> Not sure what that's about. Any thoughts?

Most likely you changed something and it can be traced to the diff between whatever you changed vs whatever is on this PR.

The CI does not get this error and neither do I. I don't get this error locally:

```
(sktree) adam2392@adams-mbp-3 scikit-tree % poetry run poe type_check
Poe => mypy -p sktree --config-file pyproject.toml
Success: no issues found in 14 source files
(sktree) adam2392@adams-mbp-3 scikit-tree % git branch
  causaltree
  cmiknn
  gh-pages
  kernel
  main
* reg-trees
  regressor-trees
  submodule
(sktree) adam2392@adams-mbp-3 scikit-tree % 
```

File: docs/references.bib [hunk 15]
```diff
@@ -29,14 +29,13 @@ @inproceedings{Meghana2019_geodesicrf
...
   booktitle = {Proceedings of the 26th ACM SIGKDD International Conference on Knowledge Discovery & Data Mining},
-  pages     = {513–523},
+  pages     = {513–-523},
```
Comment (adam2392) [inline docs/references.bib:32]: Why is this necessary?

**Comment by SUKI-O** [comment]
Can you please point me to any other unittest I need to add?

**Comment by adam2392** [comment]
I think it is mostly good now. Can you fix the circleCI issues? These are doc errors that you can check building the documentation locally.

**Comment by SUKI-O** [comment]
> I think it is mostly good now. Can you fix the circleCI issues? These are doc errors that you can check building the documentation locally.

I am getting this error when I try building docs `docstring of sktree.tree._classes.PatchObliqueDecisionTreeRegressor:154: WARNING: Explicit markup ends without a blank line; unexpected unindent.` I don't see where the it is pointing to...

File: sktree/tree/_classes.py [hunk 16]
```diff
@@ -1106,51 +1075,71 @@ class PatchObliqueDecisionTreeClassifier(DecisionTreeClassifier):
...
+    trees with the caveat that there is an increased computation. It is
+    always recommended to sample more if one is willing to spend the
+    computational resources.
+    The default values for the parameters controlling the size of the trees
+    (e.g. ``max_depth``, ``min_samples_leaf``, etc.) lead to fully grown and
+    unpruned trees which can potentially be very large on some data sets. To
+    reduce memory consumption, the complexity and size of the trees should be
+    controlled by setting those parameter values.
+    The :meth:`predict` method operates using the :func:`numpy.argmax`
+    function on the outputs of :meth:`predict_proba`. This means that in
```
Comment (adam2392) [inline sktree/tree/_classes.py]: There is no `predict_proba` method inside regressors. Overall, I would go through the documentation again to make sure it matches what semantically the class/function is trying to do.

File: sktree/tree/_classes.py [hunk 17]
```diff
@@ -1206,67 +1194,911 @@ def fit(self, X, y, sample_weight=None, check_input=True):
...
+    >>> from sklearn.datasets import load_diabetes
+    >>> from sklearn.model_selection import cross_val_score
+    >>> X, y = load_diabetes(return_X_y=True)
+    >>> from sktree.tree import PatchObliqueDecisionTreeRegressor as RGS
+    >>> regressor = RGS(random_state=0)
+    >>> cross_val_score(regressor, X, y, cv=10)
+    ...                    # doctest: +SKIP
+    ...
+    array([-0.10163671, -0.78786738,  0.01490768,  0.32737289, -0.24816698,
+            0.41881754,  0.0588273 , -1.48722913, -0.07927208, -0.15600762])
```
Comment (adam2392) [inline sktree/tree/_classes.py]: This is incorrect.

It should be
```suggestion
    References
    ----------
    .. footbibliography::
    
    Examples
    ----------
    >>> from sklearn.datasets import load_diabetes
    >>> from sklearn.model_selection import cross_val_score
    >>> X, y = load_diabetes(return_X_y=True)
    >>> from sktree.tree import PatchObliqueDecisionTreeRegressor as RGS
    >>> regressor = RGS(random_state=0)
    >>> cross_val_score(regressor, X, y, cv=10)
    ...                    # doctest: +SKIP
    ...
    array([-0.10163671, -0.78786738,  0.01490768,  0.32737289, -0.24816698,
            0.41881754,  0.0588273 , -1.48722913, -0.07927208, -0.15600762])
```

**Review by adam2392 — COMMENTED** [review]
I have pushed up a change in https://github.com/neurodata/scikit-tree/pull/72/commits/acbd87f84b26a98e52f21ac2c52b11bf815cd674 to handle some things that are hard to find. However, the other errors are due to apparent issues that I left a comment on. 

Once you fix those and go over the doc-strings again to make sure these are correct then it should be good.

File: docs/references.bib (hunk 15 above)
Comment (SUKI-O) [reply docs/references.bib:32]: All the other refs had pages listed in that format.

File: sktree/tree/_classes.py (hunk 17 above)
Comment (SUKI-O) [reply sktree/tree/_classes.py]: fixed

**Comment by adam2392** [comment]
I've fixed the documentation errors in the CI pipelines and caught some minor doc bugs.

@SUKI-O and @sampan501 lmk if the changes look good/correct to you. Please feel free to ask any questions. I'll merge when you guys hit the approve button

**Review by adam2392 — APPROVED** [review]
LGTM. Will wait for everyone to chime in given the large diff

**Comment by SUKI-O** [comment]
> LGTM. Will wait for everyone to chime in given the large diff

Thank you for all the fixes. Looks good :)

**Comment by sampan501** [comment]
Looks good to me!
